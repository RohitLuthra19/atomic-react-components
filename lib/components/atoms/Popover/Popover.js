// @flow
/**
 *
 * Popover
 *
 */
import { PureComponent, createRef } from 'react';
import styled from 'styled-components';

import styles from './Popover.style';
import type { Props, State } from './types';

class Popover extends PureComponent<Props, State> {
  static defaultProps = {
    popOverBody: '',
    popOverHeader: '',
  };

  state: State = {
    showPopover: false,
  };

  wrapperRef: { current: null | HTMLDivElement } = createRef();

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  handleClickOutside = (e: MouseEvent) => {
    const el = e.target;
    if (
      this.wrapperRef &&
      this.wrapperRef.current &&
      el instanceof Node &&
      !this.wrapperRef.current.contains(el)
    ) {
      this.setState({
        showPopover: false,
      });
    }
  };

  handleClick = () => {
    this.setState({
      showPopover: true,
    });
  };

  render() {
    const { popOverBody, popOverHeader, children } = this.props;
    const { showPopover } = this.state;
    return (
      <div role="presentation" onClick={this.handleClick} ref={this.wrapperRef}>
        {children}
        {showPopover && (
          <div role="dialog">
            <div>{popOverHeader}</div>
            <div>{popOverBody}</div>
          </div>
        )}
      </div>
    );
  }
}

export default styled(Popover)`
  ${styles};
`;

export { Popover as PopoverVanilla };
