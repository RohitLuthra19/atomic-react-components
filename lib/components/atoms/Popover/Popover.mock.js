const defaultConfig = {
  className: ``,
  popOverBody: `This is a popover body text`,
  popOverHeader: `This is a popover header text`,
};

export default defaultConfig;
