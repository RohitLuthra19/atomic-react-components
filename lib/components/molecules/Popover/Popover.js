// @flow
/**
 *
 * Popover
 *
 */
import React, { PureComponent, createRef } from 'react';
import styled from 'styled-components';

import styles from './Popover.style';
import type { Props, State } from './types';

class Popover extends PureComponent<Props, State> {
  static defaultProps = {
    popOverBody: '',
    popOverHeader: '',
    isVisible: false,
    hidePopoverCloseBtn: false,
  };

  /**
   * showPopover: State to manage the visibility of Popover content
   * default is false which can be changed by passing isVisible prop
   */
  state: State = {
    showPopover: false,
  };

  /**
   * wrapperRef: It represents the wrapper containing the element around
   * which popover needs to be wrapped and the popover content managed by
   * showPopover state or isVisible prop
   */
  wrapperRef: { current: null | HTMLDivElement } = createRef();

  /**
   * @componentDidMount
   * Add eventlistner to handle click outside the wrapper to that popover
   * gets closed and setting showPopover state on the basis of prop if passed
   */
  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
    const { isVisible } = this.props;
    this.setState({
      showPopover: isVisible,
    });
  }

  /**
   * @componentWillUnmount
   * Remove all event listener on component unmounting
   */
  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  /**
   * @handleClickOutside
   * @param {MouseEvent} e
   * This function handles the outside click from the wrapper
   * to close the popover if open
   */
  handleClickOutside = (e: MouseEvent) => {
    const el = e.target;
    if (
      this.wrapperRef &&
      this.wrapperRef.current &&
      el instanceof Node &&
      !this.wrapperRef.current.contains(el)
    ) {
      this.setState({
        showPopover: false,
      });
    }
  };

  /**
   * @handleClick
   * This function handles the click of child element (like button/anchor etc)
   * on which popover needs to be open
   */
  handleClick = () => {
    this.setState({
      showPopover: true,
    });
  };

  /**
   * @close
   * This function closes the popover on when cross button
   * inside the popover gets clicked
   */
  close = () => {
    this.setState({
      showPopover: false,
    });
  };

  render() {
    const { popOverBody, popOverHeader, children, hidePopoverCloseBtn, className } = this.props;
    const { showPopover } = this.state;

    /**
     * Cloning of element needs to be done so that handleClick function can
     * be wrapped around element without wrapping it around another element
     * or creating an extra DOM node
     */
    const element =
      children &&
      React.cloneElement(children, {
        onClick: this.handleClick,
      });
    return (
      <div role="presentation" ref={this.wrapperRef} className={className}>
        {element}
        {showPopover && (
          <div role="dialog" className="popover">
            {!hidePopoverCloseBtn && (
              <button className="popover__close" onClick={this.close}>
                X
              </button>
            )}
            <header className="popover__header">
              <h3>{popOverHeader}</h3>
            </header>
            <div className="popover__body">{popOverBody}</div>
          </div>
        )}
      </div>
    );
  }
}

export default styled(Popover)`
  ${styles};
`;

export { Popover as PopoverVanilla };
