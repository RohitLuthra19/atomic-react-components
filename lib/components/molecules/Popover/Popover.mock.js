import List from '../List';

const defaultVisiblePopover = {
  className: ``,
  popOverHeader: `This is a popover header text`,
  popOverBody: `Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.`,
  isVisible: true,
};

const hiddenPopover = {
  className: ``,
  popOverHeader: `This is a popover header text`,
  popOverBody: `Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.`,
  isVisible: false,
  wrapperBackground: `#0080FF`,
};

const ListPopover = {
  className: ``,
  popOverHeader: `This is a popover header text`,
  popOverBody: (
    <div>
      <p>This text is from a paragraph element</p>
      <List ListType="ol">
        <li>This is ordered list item, content of tab 1.</li>
        <li>This is ordered list item, content of tab 1.</li>
      </List>
      <p>This is another random text</p>
    </div>
  ),
};

const noCloseBtnPopover = {
  className: ``,
  popOverHeader: `This is a popover header text`,
  popOverBody: `Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.`,
  isVisible: false,
  wrapperBackground: `#FE7F9C`,
  hidePopoverCloseBtn: true,
};

export { defaultVisiblePopover, hiddenPopover, ListPopover, noCloseBtnPopover };
