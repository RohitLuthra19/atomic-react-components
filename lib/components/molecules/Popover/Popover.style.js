import styled from 'styled-components';

const PopoverContentWrapper = styled.div`
  position: relative;
  border: 1px solid #000000;
  padding: 5px;
  background: ${props => (props.wrapperBackground ? props.wrapperBackground : `#FFFFFF`)};
`;

const PopoverContentClose = styled.button`
  position: absolute;
  top: 7px;
  right: 7px;
  border: 0 none;
  cursor: pointer;
`;

export { PopoverContentWrapper, PopoverContentClose };
