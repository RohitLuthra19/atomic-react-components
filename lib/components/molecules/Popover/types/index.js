// @flow
import type { Node } from 'react';

export type Props = {
  popOverBody: Node | string,
  popOverHeader: Node | string,
  children: React$Element<*>,
  isVisible: boolean,
  wrapperBackground: string,
  hidePopoverCloseBtn: boolean,
};

export type State = {
  showPopover: boolean,
};
