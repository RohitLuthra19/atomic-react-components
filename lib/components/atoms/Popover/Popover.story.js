import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import defaultConfig from './Popover.mock';

// Import Styled Component to showcase variations
import Popover, { PopoverVanilla } from '.';

storiesOf('Atoms', module).addWithChapters('Popover', {
  chapters: [
    {
      sections: [
        {
          sectionFn: () => (
            <PopoverVanilla
              {...defaultConfig}
              className={`hide-default-sample ${defaultConfig.className}`}
            >
              Content
            </PopoverVanilla>
          ),
          options: {
            showSource: true,
            allowSourceToggling: true,
            showPropTables: true,
            allowPropTablesToggling: true,
          },
        },
      ],
    },
    {
      title: 'Primary Popover Variations',
      sections: [
        {
          title: 'Default',
          sectionFn: () => (
            <Popover {...defaultConfig}>
              <button>Click to Open</button>
            </Popover>
          ),
        },
      ],
    },
  ],
});
