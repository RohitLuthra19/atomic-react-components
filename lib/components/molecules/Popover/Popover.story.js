import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import {
  defaultVisiblePopover,
  hiddenPopover,
  ListPopover,
  noCloseBtnPopover,
} from './Popover.mock';

// Import Styled Component to showcase variations
import Popover, { PopoverVanilla } from '.';

storiesOf('Molecules', module).addWithChapters('Popover', {
  chapters: [
    {
      title: `Helloooooo`,
      sections: [
        {
          sectionFn: () => (
            <PopoverVanilla
              {...hiddenPopover}
              className={`hide-default-sample ${hiddenPopover.className}`}
            />
          ),
          options: {
            showSource: true,
            allowSourceToggling: true,
            showPropTables: true,
            allowPropTablesToggling: true,
          },
        },
      ],
    },
    {
      title: 'Popover Variations',
      sections: [
        {
          title: '[isVisible=true] variation',
          sectionFn: () => (
            <Popover {...defaultVisiblePopover}>
              <button>Click to Open</button>
            </Popover>
          ),
        },
        {
          title: '[isVisible=false] variation',
          sectionFn: () => (
            <Popover {...hiddenPopover}>
              <button>Click to Open</button>
            </Popover>
          ),
        },
        {
          title: 'List(Component) variation',
          sectionFn: () => (
            <Popover {...ListPopover}>
              <button>Click to Open</button>
            </Popover>
          ),
        },
        {
          title: 'Optional close button variation',
          sectionFn: () => (
            <Popover {...noCloseBtnPopover}>
              <button>Click to Open</button>
            </Popover>
          ),
        },
      ],
    },
  ],
});
