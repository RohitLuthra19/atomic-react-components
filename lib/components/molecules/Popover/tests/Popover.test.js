import { shallow } from 'enzyme';
// import renderer from 'react-test-renderer';

import Popover from '../index';

describe('<Popover />', () => {
  let PopoverComponent = '';
  beforeEach(() => {
    PopoverComponent = shallow(<Popover>Test</Popover>);
  });

  test('should render correctly', () => {
    expect(PopoverComponent).toMatchSnapshot();
  });
});
