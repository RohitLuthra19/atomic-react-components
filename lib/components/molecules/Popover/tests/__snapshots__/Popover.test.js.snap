// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popover /> should render correctly 1`] = `
<Popover
  className="Popover-sc-1xo3w5k-0 dLBzeO"
  hidePopoverCloseBtn={false}
  isVisible={false}
  popOverBody=""
  popOverHeader=""
  wrapperBackground=""
>
  Test
</Popover>
`;
